### DISCUSSION TITLE
On-demand GetServerSideProps is not working as intended

### DISCUSSION THREAD

**Discussion by the-mgi** [vercel/next.js#51351] [Help] [answered]
### Summary

on-demand getStaticProps is not working as intended, it should get the updated data per request, and if the revalidate: number is reached, it should get the data, but it is only getting data after the revalidate: number time has passed.

```javascript {.line-numbers}
function Blog({ posts }) {
  return (
    <ul>
      {posts.map((post) => (
        <li key={post.id}>{post.title}</li>
      ))}
    </ul>
  )
}
 
// This function gets called at build time on server-side.
// It may be called again, on a serverless function, if
// revalidation is enabled and a new request comes in
export async function getStaticProps() {
  const res = await fetch('https://.../posts')
  const posts = await res.json()
 
  return {
    props: {
      posts,
    },
    // Next.js will attempt to re-generate the page:
    // - When a request comes in
    // - At most once every 10 seconds
    revalidate: 10, // In seconds
  }
}
 
// This function gets called at build time on server-side.
// It may be called again, on a serverless function, if
// the path has not been generated.
export async function getStaticPaths() {
  const res = await fetch('https://.../posts')
  const posts = await res.json()
 
  // Get the paths we want to pre-render based on posts
  const paths = posts.map((post) => ({
    params: { id: post.id },
  }))
 
  // We'll pre-render only these paths at build time.
  // { fallback: 'blocking' } will server-render pages
  // on-demand if the path doesn't exist.
  return { paths, fallback: 'blocking' }
}
 
export default Blog
```

took this code snippet from https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration, and please see the line 22?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

> I think the title is wrong? You mean: On-demand GetStaticProps is not working as intended

GetServerSideProps runs on demand. Whereas, GetStaticProps, runs at build time, and it can additionally run as part of "revalidation". 

The page you linked states:

> When a request is made to a page that was pre-rendered at build time, it will initially show the cached page.
> 
> Any requests to the page after the initial request and before 10 seconds are also cached and instantaneous.
> After the 10-second window, the next request will still show the cached (stale) page
> Next.js triggers a regeneration of the page in the background.
> Once the page generates successfully, Next.js will invalidate the cache and show the updated page. If the background regeneration fails, the old page would still be unaltered.


Most notably, the regeneration is triggered after the revalidate time. 

However, you talk about on-demand revalidation, which you do by calling an endpoint, which in turn triggers the revalidate function, and which AFAIK should "ignore" the revalidate parameter, and generate a new page as you call the revalidate function, https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration#on-demand-revalidation. 

Could you clarify where you see issues please?
